src/common/snippets/src/runtime_configurator.cpp

```diff
@@ -272,9 +259,182 @@ void RuntimeConfigurator::update_data_offsets() const {
+    RuntimeConfigurator::UnifiedLoopInfoRtParams rt_params;
+    rt_params.work_amount = loop_info->get_work_amount();
+    const auto count = loop_info->get_input_count() + loop_info->get_output_count();
+    rt_params.ptr_increments.resize(count);
+    rt_params.finalization_offsets.resize(count);
+
+    size_t idx = 0;
+    loop_info->iterate_through_descs(
+        [&rt_params, &idx](const lowered::UnifiedLoopInfo::LoopPortDesc& desc) {
+            rt_params.ptr_increments[idx] = desc.ptr_increment;
+            rt_params.finalization_offsets[idx] = desc.finalization_offset;
+            ++idx;
+        });
+    return rt_params;
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::init(const lowered::LinearIRPtr& linear_ir,
+                                                    const std::vector<PortDescriptorPtr>& io_descs,
+                                                    size_t in_num) {
+    if (linear_ir->get_config().m_enable_domain_optimization || !linear_ir->is_dynamic())
+        return;
+    concurrency = linear_ir->get_config().m_min_parallel_work_amount;
+
+    const auto brgemms = find_applicable_brgemms(linear_ir);
+    // Parallel WA optimization is Brgemm related
+    if (brgemms.empty())
+        return;
+
+    init_non_m_related_params(linear_ir, brgemms);
+    OPENVINO_ASSERT(!not_m_related_params.empty(), "not_m_related_params mustn't be empty after initialization");
+    init_loops_to_split(linear_ir);
+
+    m_dim_idces.resize(io_descs.size());
+    split_layouts.resize(io_descs.size());
+    for (size_t i = 0; i < io_descs.size(); ++i) {
+        const auto& layout = io_descs[i]->get_layout();
+        const auto dim_idx = i < in_num ? utils::get_input_dim_idx(layout, 1)
+                                        : utils::get_output_dim_idx(layout, 1);
+        m_dim_idces[i] = dim_idx;
+        split_layouts[i] = SplitDimensionM::get_updated_order(layout, i < in_num ? dim_idx : layout.size() - 2);
+    }
+}
+
+bool RuntimeConfigurator::ParallelWAOptimizer::need_optimize(const VectorDims& master_shape) {
+    return !loops_to_split.empty() && SplitDimensionM::split(master_shape, concurrency, batch_m, new_m);
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::update(RuntimeConfigurator::LoopInfoRuntimeParamsMap& map,
+                                                      std::vector<VectorDims>& shapes,
+                                                      std::vector<std::vector<size_t>>& layouts,
+                                                      const std::shared_ptr<RuntimeConfig>& config,
+                                                      size_t in_num) {
+    update_split_loops_info(map);
+    update_shapes(shapes);
+    update_layouts(layouts);
+    update_config(config);
+}
+
+void RuntimeConfigurator::ParallelWAOptimizer::update_split_loops_info(LoopInfoRuntimeParamsMap& initialized_info) {
+    for (const auto& loop : loops_to_split) {
+        if (initialized_info.count(loop) == 0) {
```

Comment:
Do we really need to check that this Loop is is not in map?
`loops_to_split` is set and keys of `initialized_info` are unique. Seems like we can skip this check.
Correct me please if I'm wrong!

Reply:
Strictly speaking, `initialized_info` may be not empty, so I left this check. But it seems like the current logic implies that the map is empty before `update_split_loops_info`. So I added the corresponding assert and removed the check. Thanks!